### ISSUE TITLE
SIP Trunk with Asterisk

### ISSUE THREAD

**Issue by KTarun003** [livekit/sip#231]
Hi,

I would like my agent to accept inbound calls or make outbound calls through asterisk server. I have searched the entire documentation, but I could only find the steps Twilio or Telnyx. Any help will be much appreciated.

**Comment by dennwc** [comment]
Hi @KTarun003!

The steps to accept inbound or make outbound calls for Asterisk is the same as with Twilio or Telnyx. We do not provide documentation for it, because Asterisk can be configured in many different ways and we obviously cannot cover them all. 

Is there any specific step of the [quickstart](https://docs.livekit.io/sip/quickstarts/accepting-calls/) that doesn't work with your Asterisk setup? Also, please consider joining our [Slack](https://livekit.io/join-slack) - we usually answer general support questions there.

**Comment by KTarun003** [comment]
I am facing issue when I want to create a trunk, I dont know which URI to point the trunk. And I am using examples/voice-pipeline-agent/function_calling_weather.py example. Are there any changes to this code when using SIP. I am trying to self-host. if that helps.

**Comment by arthurblake** [comment]
I have a similar question. I just asked it on the #sip slack channel, and I referenced this issue too.

**Comment by rodrigoGA** [comment]
Great work on the project, and thanks for sharing it!
I’m trying to connect it with an Asterisk system, but for now, I’m starting with a softphone like Zoiper to test the integration, and I haven’t been able to get it working correctly.
I’m using the LiveKit API to create rules similar to these:
SIP Trunk Creation: lk sip inbound create inbound-trunk.json 
```{
  "trunk": {
    "name": "Softphone inbound trunk",
    "numbers": ["+9999999999"], 
    "auth_username": "sip-user",
    "auth_password": "sip-pass",
    "allowed_addresses": ["0.0.0.0/0"]
  }
}
```
Dispatch Rule Creation: lk sip dispatch create dispatch-rule.json  
```
{
  "name": "rodrigo-test-1 dispatch rule 3",
  "trunk_ids": ["<id>"],
  "rule": {
    "dispatchRuleIndividual": {
      "roomPrefix": "call"
    }
  }
}
```

The issue is that I can’t register the softphone with LiveKit. I’ve followed the official documentation and successfully set up an example with Twilio by associating a phone number, but now I’d like to test it with Zoiper to eventually integrate the system with Asterisk.

I’m registering the softphone with the following details:

Username: sip-user
Password: sip-pass
Domain/Host: 6o3m8e2088l.sip.livekit.cloud

Do you have any advice or additional steps to make this configuration work correctly?
Any help would be greatly appreciated!

**Comment by charly17** [comment]
Hi,
In my case, I successfully completed the integration with Asterisk with the help of fellow developers and by using LiveKit documentation. The process involves creating a dispatch rule and a room to obtain the SIP URI and registering the number within LiveKit.

You should not register the trunk outside of LiveKit for outgoing calls. For incoming calls, you only need to send the call to the SIP URI you get from LiveKit.

On the server where you have Asterisk and currently receive the DID from your provider to a standard Asterisk extension (this is to verify that everything is working fine on your provider's side and with your Asterisk server), once validated, you should route the call from your Dialplan like this:
same => n,Dial(SIP/${EXTEN}@dilivekitetcetc.sip.livekit.cloud)

As mentioned, the call must negotiate using the ulaw-alaw codecs, as I encountered issues when only G729 was being sent in my configuration.

For outgoing calls, you need to configure LiveKit so that when the server launches, it handles the call to the destination number (this is the "participant" according to LiveKit documentation). You will also need to create a trunk in Asterisk and register it in LiveKit since it will generate the call to your SIP provider connected to the PSTN. Therefore, you must configure Asterisk to manage this process (create the context where the call will land and send it to your SIP provider).

**Comment by dydimos** [comment]
Hi,
I try to get livekit also running with asterisk, but I have a complete local setup with an ATA to connect an analog phone, livekit and livekit-sip locally running in a docker container. I managed it that when I call a defined number, the phone enters the room (I use "dispatchRuleDirect"). But the phone continues to ring - the room is not answering the phone. 

Any help would be appreciated - especially @charly17 - do you maybe have an idea whats the problem on my side?
Thanks in advance :)

**Comment by dennwc** [comment]
I think one way to resolve the issues here would be if someone can provide an example of Asterisk configuration inside a Docker container.

We can then combine it with our [existing Docker compose](https://github.com/livekit/sip/blob/main/docker-compose.yaml) example for self-hosting and properly document integration with Asterisk in our official docs.

Unfortunately I don't have spare cycles right now to do it, and I'm not very familiar with Asterisk. Can someone help by making a repo/PR with an example config? That would be really helpful!

**Comment by dtcgroup** [comment]
> Hi, I try to get livekit also running with asterisk, but I have a complete local setup with an ATA to connect an analog phone, livekit and livekit-sip locally running in a docker container. I managed it that when I call a defined number, the phone enters the room (I use "dispatchRuleDirect"). But the phone continues to ring - the room is not answering the phone.
> 
> Any help would be appreciated - especially @charly17 - do you maybe have an idea whats the problem on my side? Thanks in advance :)

I know it is not an answer, but I am having exactly the same problem using FreeSwitch.

It sends the Invite, gets a Processing(100) response and then a Ringing (180) response repeated every second until the call times out with no answer.

My node application detects the call and logs the connecting to agent which suggests the dispatch and numbers all line up. We just never get the call answered by the room as you say.

**Comment by dennwc** [comment]
If you get continuous 180 responses, this means LiveKit SIP is waiting for any media from an Agent.

Due to the way SFU works currently, publishing tracks is not enough. You might need to send at least one media packet (could be empty/silence). This should unblock SIP and it will start bridging the call.

Are you seeing `Waiting for track subscription(s)` each time the call gets stuck? If so, this is it.

**Comment by dydimos** [comment]
Somehow I was able to solve it, now the call is transmitted to the room and I can hear myself speak. I used your docker-compose as a reference. 

Unfortunately now I have a next problem - see here: #200 (https://github.com/livekit/sip/issues/200#issuecomment-2552433308)

If everything works for me, I will document the setup and can post it here if there is any interest.

**Comment by umairmehar** [comment]
Hello @dydimos, I'm trying to achieve the same. 

I have a complete local setup with an ATA (Grandstream H802) to connect an analog phone, livekit, and livekit-sip locally running. 

The ATA modem is simply not registering/communicating with the local livekit-sip server. 

I confirmed that all my servers, ATA modem, livekit, livekit-sip, and Redis, are running and accessible on the local network. 

Any help would be appreciated. 

Thanks in advance.

**Comment by duzaq** [comment]
> Estou tentando conectá-lo a um sistema Asterisk, mas, por enquanto, estou começando com um softphone como o Zoiper para testar a integração, e não consegui fazê-lo funcionar corretamente. Estou usando a API do LiveKit para criar regras semelhantes a estas: Criação de tronco SIP: lk sip inbound create inbound-trunk.json
> 
> ```
>   "trunk": {
>     "name": "Softphone inbound trunk",
>     "numbers": ["+9999999999"], 
>     "auth_username": "sip-user",
>     "auth_password": "sip-pass",
>     "allowed_addresses": ["0.0.0.0/0"]
>   }
> }
> ```
> 
> Criação de regra de despacho: lk sip dispatch create dispatch-rule.json
> 
> ```
> {
>   "name": "test-1 dispatch rule softphone",
>   "trunk_ids": ["<id>"],
>   "rule": {
>     "dispatchRuleIndividual": {
>       "roomPrefix": "call"
>     }
>   }
> }
> ```
> 
> O problema é que não consigo registrar o softphone com o LiveKit. Eu segui a documentação oficial e configurei com sucesso um exemplo com o Twilio associando um número de telefone, mas agora eu gostaria de testá-lo com o Zoiper para eventualmente integrar o sistema com o Asterisk.
> 
> Estou registrando o softphone com os seguintes detalhes:
> 
> Nome de usuário: sip-user Senha: sip-pass Sip uri: 6o3m8e2088l.sip.livekit.cloud
> 
> Você tem algum conselho ou etapas adicionais para fazer essa configuração funcionar corretamente? Qualquer ajuda seria muito apreciada!

Alguma solução?

**Comment by UlfWil** [comment]
> Somehow I was able to solve it, now the call is transmitted to the room and I can hear myself speak. I used your docker-compose as a reference.
> 
> If everything works for me, I will document the setup and can post it here if there is any interest.

Of course, I'm interested in seeing your setup. Do you use livekit.cloud, or are you running a local LiveKit server?

**Comment by UmairShah7677** [comment]
Hi everyone,  

I'm attempting to integrate my Asterisk server with LiveKit SIP **(Also mentioned this in slack community)**, and I'm running into an issue where calls to the SIP URI are failing with a **503 "Try again later"** response. Below are the details of what I've done so far:  

### **Setup Overview:**
- SIP URI: `sip:2vbic7bb4fd.sip.livekit.cloud`
- Asterisk is configured to forward calls to this URI when the DID is dialed.
- The DID is confirmed to be working correctly and is properly routed to the context in Asterisk.

### **Asterisk Dialplan:**
Here’s the relevant part of my dialplan:  
```ini
exten => 9999,1,Dial(SIP/2vbic7bb4fd.sip.livekit.cloud)
exten => 9999,n,GotoIf($["${DIALSTATUS}"="CONGESTION"]?retry:hangup)
exten => 9999,n(retry),Wait(5)
exten => 9999,n,Dial(SIP/2vbic7bb4fd.sip.livekit.cloud)
exten => 9999,n(hangup),Hangup()
```

### **Issue:**
When I call the DID, Asterisk attempts to route the call to LiveKit, but I receive the following logs:  
```
-- Called SIP/2vbic7bb4fd.sip.livekit.cloud
-- Got SIP response 503 "Try again later" back from 84.235.238.224:5060
-- SIP/2vbic7bb4fd.sip.livekit.cloud-0000002d is circuit-busy
== Everyone is busy/congested at this time (1:0/1/0)
```
This results in a `CONGESTION` status and retries as per the dialplan logic. However, the issue persists even after retries.

### **Steps Taken to Troubleshoot:**
1. **Codec Compatibility**: I ensured that only `ulaw` and `alaw` codecs are allowed in both Asterisk and LiveKit.
2. **SIP URI Verification**: Confirmed the SIP URI is correct and matches the one provided by LiveKit.
3. **Firewall**: Verified that there are no firewall restrictions blocking traffic to/from LiveKit (`84.235.238.224:5060`).
4. **LiveKit Dispatch Rules**: Checked that the dispatch rule routes calls to the correct room in LiveKit.
5. **Connectivity**: Tested connectivity to the LiveKit server using `ping` and `telnet` for SIP port 5060.

### **Logs from Asterisk**:
Here’s an excerpt from the logs:
```
Got SIP response 503 "Try again later" back from 84.235.238.224:5060
SIP/2vbic7bb4fd.sip.livekit.cloud-0000002d is circuit-busy
```

### **Request for Help:**
- Is the **503 "Try again later"** response indicative of an issue on LiveKit's side, or could it be related to my configuration?
- Are there additional debugging steps I can take to identify the cause of this issue?
- Could this be related to resource availability or misconfigured dispatch rules in LiveKit?

Any guidance or suggestions would be greatly appreciated. Thank you in advance for your time and help!  

Best regards.

**Comment by MefhigosetH** [comment]
Hi @UmairShah7677,

I don't see that Asterisk is sending the extension number when Dial. You can try this: `Dial(SIP/EXT_NUMBER@2vbic7bb4fd.sip.livekit.cloud)` where `EXT_NUMBER` is the number that you configured in Livekit Cloud when created the Inbound Trunk.

Remember that you need configure two things in Livekit: an Inbound Trunk, and a Dispatch Rule. As documentation says.

**Comment by ccameroVzy** [comment]
Hi everyone!

We are having a very similar issue. 

We are using the cloud sip server, and first we connected our agent using sip-trunk and dispatch rules to a Twilio Number and it works just fine.

But we need to integrate with Kazoo our VoIP, so we created an extension on kazoo to LiveKit, created the SIP Trunk and dispatch rules. But in this case when we make the call, its received by LiveKit, the agent is actually dispatched to the room and we can see even the agents starts the speech 

`2025-01-20 14:14:05,947 - DEBUG livekit.agents.pipeline - speech playout started
 {"speech_id": "01fa05decd9a"}`

but on the user ends theres still ringing, and we just keep getting a 180 code and never de 200 status

![Image](https://github.com/user-attachments/assets/b38e35f9-1342-4da9-94f2-785bc53282a3)

Any guidance is helpful

**Comment by MefhigosetH** [comment]
Hi @ccameroVzy,

I've read that Livekit Cloud, only works with TCP packets. Not UDP. Try with TCP communication.

**Comment by ccameroVzy** [comment]
@MefhigosetH yes, that was the issue.

Now is working properly thanks!

**Comment by UmairShah7677** [comment]
Hey @MefhigosetH Thank you for help, it is connected  but now it is keep ringing. Can you please tell me @ccameroVzy that how you fixed as after this ```Dial(SIP/EXT_NUMBER@2vbic7bb4fd.sip.livekit.cloud)```
the 503 error is solved, but its keep on ringing, Also i can see session on the livekit cloud. 

![Image](https://github.com/user-attachments/assets/13ed7e4b-c200-4f78-ba4b-0704fa99d0e1)

In ```sngrep``` it is giving response of "200" but i can't hear any agent.

![Image](https://github.com/user-attachments/assets/73a2e0bf-235b-4d4e-b8f6-704d28f0dd42)

**Comment by UmairShah7677** [comment]
Thank you @MefhigosetH working fine.

**Comment by PatroxGaurab** [comment]
We enabled the TCP as well, but it keeps ringing
while livekit agent is publishing track on the room properly as per livekit agents logs

**Comment by stiven77nj** [comment]
Me está sucediendo igual. Se realiza el registro de la llamada, pero no hay una respuesta. No se si estoy haciendo algo mal

![Image](https://github.com/user-attachments/assets/f37e31b9-af35-422f-8c75-7eed1b42f538)

Paso a paso de lo que me ha sucedido:

- Paso 1:  Me aparecía el siguiente error:

```Called SIP/2vbic7bb4fd.sip.livekit.cloud
Got SIP response 503 "Try again later" back from 84.235.238.224:5060
SIP/2vbic7bb4fd.sip.livekit.cloud-0000002d is circuit-busy
Everyone is busy/congested at this time (1:0/1/0)
```

Lo pude solucionar agregando lo siguiente en el archivo de **Asterisk**:

```Dial(SIP/{EXT_NUMBER}@2vbic7bb4fd.sip.livekit.cloud)```

- Paso 2: Ahora al realizar la llamada se queda sonando, hasta colgarse por sí sola y es ahi donde aparece registro de la llamada en el dashboard de LiveKit

![Image](https://github.com/user-attachments/assets/40a226c2-c971-48f7-b196-c0c50cae2b4c)

Agradezco cualquier ayuda... Gracias!!!

**Comment by MefhigosetH** [comment]
Hi.

I think a VoIP traffic trace would be helpful. Maybe a sngrep trace?

I understand that the call will keep ringing until the first DSP (handshake) packet arrives 🤷🏻‍♂️

**Comment by stiven77nj** [comment]
¿Esta configuración es correcta?

```type=peer
host=#####.sip.livekit.cloud
port=5060
context=from-trunk-custom
insecure=port,invite
nat=yes
qualify=yes
dtmfmode=rfc2833
disallow=all
allow=opus
canreinvite=no
transport=tcp
```

¡¡¡Agradezco cualquier ayuda... Gracias!!!

**Comment by stiven77nj** [comment]
Hola, estuve revisando lo que me dices, pero no vi nada claro. Sin embargo, hizo un montaje en local con docker como la siguiendo la documentación. Estoy relizando unas llamadas y los logs de mis contenedores aparece lo siguiente:

- servidor liveKit:

![Image](https://github.com/user-attachments/assets/b0a018f6-1636-4b28-8f53-02c51b35064e)

- servidor sip:

![Image](https://github.com/user-attachments/assets/40b272d3-f659-4627-86ec-34e6d93b2934)

Alguna idea??... Ayuda por favor

**Comment by MefhigosetH** [comment]
Hola @stiven77nj 👋🏻 

Por el mensaje del servidor/contenedor livekit, pareciera que no existe un inbound trunk que coincida con la llamada entrante que le estás enviando. También debes crear una Dispatch Rule.

Sería el paso 3 de este instructivo: https://docs.livekit.io/agents/quickstarts/inbound-calls/

**Comment by stiven77nj** [comment]
Hola @MefhigosetH !

La verdad, no entiendo. Sigue sin funcionar.

Este es el paso a paso de lo hice:

- .env.local

![Image](https://github.com/user-attachments/assets/4282ff55-63b5-48d0-ad31-0e29c2ad42d1)

- inbound-trunk.json
 
![Image](https://github.com/user-attachments/assets/8fdd425a-a1be-49dc-b625-34e1d047add8)

- dispatch-rule.json

![Image](https://github.com/user-attachments/assets/ec759f8d-84d6-4609-9fa0-6dee058d5280)

Los IDs de cada archivo los genero de la siguiente manera:

```
lk sip inbound create inbound-trunk.json
```

```
lk sip dispatch create dispatch-rule.json
```

No se si exista una guia o alguien que haya podido realizar la conexion de manera exitosa, me pueda ayudar!.

**Gracias!!**

**Comment by UmairShah7677** [comment]
Hi, is there any solution for ringing, As when the call is routing towards self hosted sip server, it keeps on ringing, i turned off the sounds on my own asterisk server, but this ringing is coming from the livekit sip server, how can i turn the ring off.

**Comment by erwiower2** [comment]
> Hi, is there any solution for ringing, As when the call is routing towards self hosted sip server, it keeps on ringing, i turned off the sounds on my own asterisk server, but this ringing is coming from the livekit sip server, how can i turn the ring off.

Use:
same => n,Dial(SIP/${EXTEN},**,m**)

The **m,** option is important to suppress dialtone. 

**Make sure the following steps because m option starts your MOH:**

> Do not use the [default] context in your musiconhold.conf if your MOH starts playing automatically.

Leaving the [default] context defined can lead to unexpected behavior. In particular, when a remote party puts you on hold, Asterisk may start playing its own Music on Hold (MOH) instead of using the one provided by the other party.

This can result in warnings like:
WARNING[2020][C-0000003d]: res_musiconhold.c:971 _get_mohbyname: Music on Hold class ‘default’ not found in memory. Verify your configuration.

To avoid this, it’s best practice to define a separate context for each MOH class in musiconhold.conf and explicitly start it in your dialplan using:
StartMusicOnHold(YourMOH)

Note that StartMusicOnHold() no longer works without a defined context, and Asterisk will not fall back to a default context anymore.

**Comment by UmairShah7677** [comment]
Perfect!. I used this n,Dial(SIP/${EXTEN},,m) and fine.
Thank you @erwiower2

**Comment by attiquers** [comment]
@erwiower2 @UmairShah7677 
can you please give me the your asterisk and livekit configuration here so i can see what i did wrong on my side. Or Can you tell me what i am doing wrong. Call connects but i dont hear audio or cant send audio on zoiper.  
I am using docker livekit and local asterisk.


Here’s your configuration and Python code formatted nicely in Markdown:

```ini
====================== pjsip.conf =====================
[transport-tcp]
type=transport
protocol=tcp
bind=0.0.0.0

[transport-udp]
type=transport
protocol=udp
bind=0.0.0.0

[1000]
type=endpoint
context=from-internal
disallow=all
allow=ulaw,alaw,opus
auth=auth1000
aors=1000
transport=transport-tcp
media_use_received_transport=yes
rtp_symmetric=yes
timers=yes

[1000]
type=aor
max_contacts=100

[auth1000]
type=auth
auth_type=userpass
password=1000p
username=1000

[livekit]
type=endpoint
context=from-livekit
disallow=all
allow=ulaw,alaw,opus
aors=livekit
transport=transport-tcp
media_use_received_transport=yes
rtp_symmetric=yes

[livekit]
type=aor
contact=sip:127.0.0.1:5070
```

```ini
====================== extensions.conf =====================
[general]
static=yes
writeprotect=no

[from-internal]
exten => _X.,1,NoOp(Calling ${EXTEN} via LiveKit)
same => n,Dial(PJSIP/${EXTEN}@livekit,60)
same => n,Hangup()

[from-livekit]
exten => _X.,1,NoOp(Incoming call from LiveKit to ${EXTEN})
same => n,Answer()
same => n,Dial(PJSIP/1000,60)
same => n,Hangup()
```

```ini
====================== rtp.conf =====================
[general]
rtpstart=10000
rtpend=20000
icesupport=yes
rtpkeepalive=10
```

```json
====================== inbound.json =====================
{
  "trunk": {
    "name": "My trunk",
    "numbers": ["+15105550100"],
    "krispEnabled": true
  }
}
```

```json
====================== dispatch.json =====================
{
  "dispatch_rule": {
    "rule": {
      "dispatchRuleIndividual": {
        "roomPrefix": "call-"
      }
    },
    "name": "My dispatch rule",
    "roomConfig": {
      "agents": [
        {
          "agentName": "inbound-agent",
          "metadata": "job dispatch metadata"
        }
      ]
    }
  }
}
```

```python
====================== agent.py =====================
from livekit import agents
from livekit.agents import AgentSession, Agent
from livekit.plugins import openai
import asyncio
from dotenv import load_dotenv

load_dotenv()

class Assistant(Agent):
    def __init__(self) -> None:
        super().__init__(instructions="You are a helpful voice AI assistant.")

async def entrypoint(ctx: agents.JobContext):
    session = AgentSession(
        llm=openai.realtime.RealtimeModel(voice="coral")
    )

    await session.start(room=ctx.room, agent=Assistant())

    # Send a greeting once
    await session.generate_reply(instructions="Welcome to LiveKit voice assistant.")
    print("Greeting sent! Audio should reach Zoiper.")

    # Keep the session alive so Zoiper can hear it
    await asyncio.sleep(300)  # 5 minutes

if __name__ == "__main__":
    agents.cli.run_app(
        agents.WorkerOptions(
            entrypoint_fnc=entrypoint,
            agent_name="inbound-agent"
        )
    )
```

**Comment by MefhigosetH** [comment]
Hi @attiquers ,

Krisp_enabled must be used only when you use Livekit Cloud. Try disable this setting.

Additionally, try see Livekit SIP Docker logs, so you can catch any error. Maybe you need configure Livekit SIP to not try resolve public IP and use local IP while SDP handshake.

**Comment by erwiower2** [comment]
I found out that no running agent.py is required. I started the script and then stopped it again. Before I did command _lk cloud auth_ in my terminal. I am new to LiveKit and my project is not ready for production. I still have to figure out a few things.


**_Be careful when copying and keep in mind that individual adjustments may be necessary!_**

I am using chan_sip, so don't forget to change in extensions.conf _Dial(SIP/...)_ to _Dial(PJSIP/...)_

sip.conf:
```
[livekit]
type=peer
host=xxx.sip.livekit.cloud
port=5060
insecure=port,invite
disallow=all
allow=alaw ;possible ulaw, alaw, opus, g722. I prefer alaw for Europe
context=outgoing
dtmfmode=rfc2833
```

extensions.conf:
```
[outgoing]
exten => 7001,1,Answer()
same => n,Dial(SIP/7001@livekit,,m) ;<- Make sure that is your inbound number on your livekit dashboard
same => n,Hangup()

exten => 7002,1,Answer()
same => n,Dial(SIP/7002@livekit,,m) ;<- Make sure that is your inbound number on your livekit dashboard
same => n,Hangup()
```

agent.py:
```
#!/etc/asterisk/scripts/livekit/xxx/venv/bin/python3.9
import logging
import os
from livekit.plugins import noise_cancellation
from dotenv import load_dotenv
from livekit.agents import (
    Agent,
    AgentSession,
    AutoSubscribe,
    JobContext,
    JobProcess,
    WorkerOptions,
    cli,
    metrics,
    RoomInputOptions,
)

load_dotenv(dotenv_path="/etc/asterisk/scripts/livekit/xxx/.env.local")
logger = logging.getLogger("voice-agent")
os.environ['SSL_CERT_FILE'] = '/etc/ssl/certs/ca-certificates.crt'
os.environ['SSL_CERT_DIR'] = '/etc/ssl/certs/'

class Assistant(Agent):
    def __init__(self) -> None:
        super().__init__(
            instructions="This agent is a simple bridge for LiveKit voice rooms.",
        )

    async def on_enter(self):

        pass


def prewarm(proc: JobProcess):
    pass


async def entrypoint(ctx: JobContext):
    logger.info(f"connecting to room {ctx.room.name}")
    await ctx.connect(auto_subscribe=AutoSubscribe.AUDIO_ONLY)

    logger.info(f"starting voice assistant session in room {ctx.room.name}")


    usage_collector = metrics.UsageCollector()

    def on_metrics_collected(agent_metrics: metrics.AgentMetrics):
        metrics.log_metrics(agent_metrics)
        usage_collector.collect(agent_metrics)

    session = AgentSession()

    session.on("metrics_collected", on_metrics_collected)

    await session.start(
        room=ctx.room,
        agent=Assistant(),
        room_input_options=RoomInputOptions(
            # noise_cancellation=noise_cancellation.BVC(),
        ),
    )


if __name__ == "__main__":
    cli.run_app(
        WorkerOptions(
            entrypoint_fnc=entrypoint,
            prewarm_fnc=prewarm,
        ),
    )
```

dispatch-rule.json:
```
{
  "sipDispatchRuleId": "SDR_ID",
  "rule": {
    "dispatchRuleDirect": {
      "roomName": "my_room"
    }
  },
  "trunkIds": [
    "ST_TRUNKID1",
    "ST_TRUNKID2"
  ],
  "name": "Asterisk Dispatch",
  "roomConfig": {}
}
```

Then run command:
```
lk sip dispatch create dispatch-rule.json
```

**However, I would recommend making all these settings related to LiveKit through the LiveKit web interface, depending on what is easier: Telephony > Configuration.**


You will run into some further issues, such as:

- How to avoid having the MoH (Music on Hold) go through LiveKit.
You will therefore need to monitor Hold/Unhold events via scripts and, depending on the condition, “artificially” put the existing call on hold on another local channel, so that the remote party is no longer communicating through your LiveKit room. Otherwise, the remote party would hear your MoH filtered through LiveKit and notice artifacts in the music/announcement. Forwarding calls would also no longer be straightforward, since you probably don’t want your forwarded call to continue running through LiveKit. This would consume your LiveKit call minutes. You need to imagine that both you and your call partner count as two participants in the LiveKit room. That means:

_1000 call minutes / month : 2 participants = 500 call minutes / month_


[See also](https://community.asterisk.org/t/avoid-executing-h-extension-on-ami-redirect/109506/5?u=dht768)

- How to make outgoing calls via LiveKit.
This again depends on your Asterisk configuration. In my case, it was necessary that when making an outgoing call, I triggered a call to an extension, which then accepted the number, passed it to a shell script, which invited the participant into the room. On the active extension, you then joined the room yourself via a Dial command!

extensions.conf:
```
[outgoing]
exten => _X.,1,Answer()
same => n,System(/etc/asterisk/scripts/livekit/invite_participant.sh ${EXTEN}) ;<- Your LiveKit Outbound Trunk number e.g. 7000
same => n,Dial(SIP/7002@livekit,,m) ;<- Your LiveKit Inbound Trunk number e.g. 7001 or 7002
```

invite_participant.sh:
```
#!/bin/bash

export LIVEKIT_URL="wss://xxx.cloud" 
export LIVEKIT_API_KEY="xxx"
export LIVEKIT_API_SECRET="xxx"

if [ -z "$1" ]; then
  exit 1
fi

TARGET_NUMBER="$1"
SOURCE_JSON_FILE="/etc/asterisk/scripts/livekit/callfile/outbound/sip-participant.json"
TEMP_JSON_FILE="/etc/asterisk/scripts/livekit/callfile/outbound/tmp/sip_participant_temp_$(date +%s%N).json"

if ! jq ".sip_call_to = \"$TARGET_NUMBER\"" "$SOURCE_JSON_FILE" > "$TEMP_JSON_FILE"; then
  rm -f "$TEMP_JSON_FILE"
  exit 1
fi

lk sip participant create "$TEMP_JSON_FILE"

if [ $? -eq 0 ]; then
  echo "Calling: $TARGET_NUMBER"
else
  echo "Call failed."
fi
rm -f "$TEMP_JSON_FILE"
```

sip-participant.json:
```
{
  "sip_trunk_id": "ST_OUTBOUNDTRUNKID",
  "room_name": "my_room",
  "participant_identity": "Livekit-Out",
  "participant_name": "Outgoing Call",
  "krisp_enabled": false,
  "wait_until_answered": false,
  "play_dialtone": true,
  "sip_call_to": "PLACEHOLDER"
}
```
**Also make sure in the sip-participant.json that everything is configured correctly, such as the Outbound Trunk ID, and that kisp_enabled remains set to false, because you don’t want to hear the participant filtered yourself! Only you should be heard with filtering, which is why in the LiveKit Cloud Dashboard you must make sure to check “Krisp enabled” on the Inbound Trunk!**

**Comment by attiquers** [comment]
@MefhigosetH @erwiower2 
Thank you so much for replying. Can you help me with this issue now?
I had too much issues with local so i have gone cloud now. 

When i place a call with transport-udp in livekit-cloud endpoint i see 180 ringing continuous in asterisk-cli
To cater that i switch to transport-tcp and it wont ring at all.
This code  worked last night my calls were accepted by agent, when i wokeup next morning it wont work, maybe i configured something wrong

pjsip.conf
```
[general]
static=yes
writeprotect=no
autofallthrough=yes

; Transport(s)
[transport-udp]
type=transport
protocol=udp
bind=0.0.0.0:5060
local_net=192.168.XXX.0/24
external_signaling_address=XXX
external_media_address=XXX

[transport-tcp]
type=transport
protocol=tcp
bind=0.0.0.0:5060
local_net=192.168.XXX.0/24
external_signaling_address=XXX

; ---------------------------
; LOCAL PHONE (Zoiper) - user 1000
; ---------------------------
[auth1000]
type=auth
auth_type=userpass
username=1000
password=XXX

[1000]
type=aor
max_contacts=5
remove_existing=yes

[1000]
type=endpoint
context=from-internal
disallow=all
allow=ulaw,opus
aors=1000
auth=auth1000
direct_media=no
rtp_symmetric=yes
force_rport=yes
rewrite_contact=yes
transport=transport-udp ; allow UDP by default; TCP calls will also match via identify if needed

[identify-1000]
type=identify
endpoint=1000
match=192.168.XXX.0/24

; ---------------------------
; LIVEKIT TRUNK (use as outbound trunk)
; ---------------------------
[livekit-outbound-auth]
type=auth
auth_type=userpass
username=1000
password=XXX

[registration_livekit]
type=registration
outbound_auth=livekit-outbound-auth
server_uri=sip:XXX.sip.livekit.cloud
client_uri=sip:1000@XXX.sip.livekit.cloud
contact_user=1000
retry_interval=60

[livekit-cloud-aor]
type=aor
; no static contact — we will use outbound_proxy/registration to reach LiveKit

[livekit-cloud]
type=endpoint
context=from-livekit
disallow=all
allow=ulaw,opus
outbound_auth=livekit-outbound-auth
transport=transport-tcp      ; << changed to TCP
direct_media=no
rtp_symmetric=yes
force_rport=yes
rewrite_contact=yes
outbound_proxy=sip:XXX.sip.livekit.cloud
```

extensions.conf
```
[general]
static=yes
writeprotect=no
autofallthrough=yes

[from-internal]
; Any numeric dialed from 1000 will go to LiveKit trunk
exten => _X.,1,NoOp(Forwarding ${EXTEN} to LiveKit via endpoint livekit-cloud)
 same => n,Set(LIVEKIT_NUMBER=+XXX)
 same => n,NoOp(Calling LiveKit number ${LIVEKIT_NUMBER} for dialed ${EXTEN})
 ; Dial using endpoint livekit-cloud and explicit SIP URI resource
 same => n,Dial(PJSIP/livekit-cloud/sip:${LIVEKIT_NUMBER}@XXX.sip.livekit.cloud,60)
 same => n,Hangup()

[from-livekit]
; Incoming calls from LiveKit will land here
exten => _X.,1,NoOp(Inbound call from LiveKit: ${CALLERID(num)})
 same => n,Answer()
 same => n,Playback(hello-world)   ; replace with your desired handling (AGI, queue, bridge)
 same => n,Hangup()
```

**Comment by erwiower2** [comment]
Use chan_sip > sip.conf.
chan_pjsip didn't work for me.
I am running chan_sip and chan_pjsip at the same time.
chan_sip > UDP port 5060
chan_pjsip > UDP port 5090 (or similar).

**Comment by MefhigosetH** [comment]
It's true. My instalation works with chan_sip too.

**Comment by cror07** [comment]
greetings,

I have a similar problem and is that the connection is not completed with status 200, I throw a 180 continuous, but in the development mode logs, I throw this last log:

`
2025-08-19 00:52:27,694 - DEBUG livekit.agents - committed agent speech {"agent_transcript": "Hola, gracias por llamar. \u00bfEn qu\u00e9 puedo ayudarte hoy?", "interrupted": false, "pid": 371999, "job_id": "AJ_VHsm7nfEUmDv"}
`

but the call rings continuously but at the end the call is not answered and goes to voicemail.

I also use udp and chan_sip, because the other two do not work for me.

I appreciate the help, I have been unable to solve it for 2 days, thank you.

**Comment by UmairShah7677** [comment]
Hey @cror07 
Do you have any firewall on your end, try to turn the firewall off because your agent and sip server are not on same machine if you are using self hosted livekit sip server. Also use TCP instead of UDP, i used sip on mod TCP and it worked.

**Comment by ipslhr22-blip** [comment]
i am using asterisk20 and i want to connect with livekit that my livekit app can call through my own asterisk server i cofigured it like this but getting error i am using sip.conf and all required prts are allowed in ufw no nating no firewall

[general]
context=default
externip=my live ip here
udpbindaddr=0.0.0.0:5060
allowguest=yes
tcpenable=no
udpbindaddr=0.0.0.0
transport=udp
nat=force_rport,comedia
qualify=yes
allowguest=yes
directmedia=no
icesupport=yes
localnet=10.99.0.0/16
rtpstart=10000
rtpend=20000
nat=yes
register => mytrunk:1FCba321@xxxx.sip.livekit.cloud

[livekit-trunk]
type=peer
host=xxxx.sip.livekit.cloud
defaultuser=mytrunk
secret=1FCba321
context=from-livekit
insecure=invite,port
disallow=all
allow=ulaw
qualify=yes
nat=force_rport,comedia
directmedia=no
dtmfmode=rfc2833

issue : 

livekit-trunk/mytrunk     129.151.44.211                              Yes        Yes            5060     UNREACHABLE
5 sip peers [Monitored: 2 online, 3 offline Unmonitored: 0 online, 0 offline]
[Sep 16 22:39:12] NOTICE[44910]: chan_sip.c:16062 sip_reg_timeout:    -- Registration for 'mytrunk@xxxx.sip.livekit.cloud' timed out, trying again (Attempt #53)

can anyone help me regarding 

i added outbound trunk in live kit : 

{
  "sipTrunkId": "ST_5EjXfu6BRbxx",
  "name": "My outbound trunk",
  "address": "my live ip here",
  "numbers": [
    "1001"
  ],
  "authUsername": "mytrunk",
  "authPassword": "1FCba321"
}

inbound : 
{
  "sipTrunkId": "ST_FDsUNHg3ogg9",
  "name": "myTrunk",
  "numbers": [
    "1000"
  ]
}

at this time i want to call at my locall extension from my livekit project

**Comment by MefhigosetH** [comment]
Hi @ipslhr22-blip ,

Try with SIP over TCP.

**Comment by erwiower2** [comment]
Try this:

```
[general]
context=default
bindport=5060
bindaddr=0.0.0.0
tcpenable=yes
transport=tcp
srvlookup=yes
dtmfmode=auto
disallow=all
allow=ulaw


[livekit-trunk]
type=peer
host=xxxx.sip.livekit.cloud
port=5060
insecure=port,invite
disallow=all
allow=ulaw
context=default

```
**And please do me a favor: delete your last post and change all your passwords! You have exposed sensitive information!**
@ipslhr22-blip

**Comment by arthurblake** [comment]
I just wanted to refute the earlier comments that implied there might be a problem with PJSIP and that chan_sip might be the key.
Everything works fine for me with asterisk and LiveKit on both chan_sip and PJSIP. PJSIP is a much more robust stack (chan_sip is no longer even an option in the latest versions of asterisk, so that's not even an option anymore since we updated to the latest Asterisk 22 LTS)

**Comment by singh-harveer** [comment]
HI Everyone,
I am getting `media timeout` error in livekit cloud call logs.

I have install asterisk on ec2 with below configuration:


 -----pjsip.conf------

```[transport-udp]
type=transport
protocol=udp
bind=0.0.0.0

[transport-tcp]
type=transport
protocol=tcp
bind=0.0.0.0


local_net=EC2_PRIVATE_IP/32
external_media_address=XX.XX.XX.XX  # ec2 public ip
external_signaling_address=XX.XX.XX.XX # # ec2 public ip

[110000]
type=endpoint
aors=110000
auth=110000
context=from-internal
disallow=all
allow=ulaw,alaw
rtp_symmetric=yes
force_rport=yes
direct_media=no

[110000]
type=aor
max_contacts=1
remove_existing=yes

[110000]
type=auth
auth_type=userpass
password=xccvbxxx
username=110000

[200000]
type=endpoint
aors=200000
auth=200000
context=from-internal
disallow=all
allow=ulaw,alaw
rtp_symmetric=yes
force_rport=yes
direct_media=no

[200000]
type=aor
max_contacts=1
remove_existing=yes

[200000]
type=auth
auth_type=userpass
password=somepassword
username=200000

[LiveKit-Trunk]
type=endpoint
transport=transport-tcp
context=from-internal
disallow=all
allow=ulaw,alaw
aors=LiveKit-AOR
rtp_symmetric=yes
force_rport=yes
direct_media=no

[LiveKit-AOR]
type=aor
contact=sip:xxxxx.sip.livekit.cloud
```


--------rtp.conf---------

```
; RTP Configuration
[general]
rtpstart=10000
rtpend=20000
strictrtp=no
rtp_keepalive=1
```

---- secrutiy group config------

```
|  IP Version  |  Rule Type   |  Protocol  |  Port/Port Range  |  Source     |  Description  |  Target
------------------------+--------------+--------------+------------+-------------------+-------------+---------------+--------
|  IPv4        |  Custom TCP  |  TCP       |  5060             |  0.0.0.0/0  |  –            |  –     
|  IPv4        |  Custom UDP  |  UDP       |  10000 - 20000    |  0.0.0.0/0  |  –            |  –     
|  IPv4        |  Custom UDP  |  UDP       |  5060             |  0.0.0.0/0  |  –            |  –     
|  IPv4        |  SSH         |  TCP       |  22               |  0.0.0.0/0  |  –            |  –     


----- extentions.conf------

```[from-internal]
exten => 1000,1,Dial(PJSIP/+177541XXX@LiveKit-Trunk) # +177541XXX this number registered in livekit trunk and working, we have tested it with twilio
exten => 1000,2,Hangup()
```


We are using livekit cloud for sip and media server.

Livkit agent is running in my local laptop, which is registering it self to livekit cloud

Now with this config I can see my voice/audio is reaching to livekit agent, but I am not able to hear back from agent.

I can see agent is speaking as well from generated log, but can't hear audio  back.
In Livekit cloud call logs I can see `media timeout` error.

I am struggling with it for long time,Any help will be appriciated.

thanks!

**Comment by attiquers** [comment]
Alright boys, i have done it. Below is the full guide for latest Asterisk-22. It goes like this:
1. Installing Asterisk (I am using azure cuz free student credits go brrrr) 
2. Copying the configuration files to your Asterisk
3. Opening essential ports so asterisk and livekit can communicate properly
4. Creating inbound-trunk and dispatch rule on LivekitCloud (or selfhost)
5. Simple agent code, so it joins the room when you call
6. Some common issues you might run into and fixes 
7. Optional: Shameless plug to hire my unemployed ass


# 1. Installing Asterisk
**You must have atleast 1GB RAM on your VM otherwise it crashes mid-installation**

### Install dependencies Asterisk needs
```
sudo apt-get update

sudo apt install bison wget openssl libssl-dev libasound2-dev libc6-dev libxml2-dev libsqlite3-dev libnewt-dev libncurses-dev zlib1g-dev gcc g++ make perl uuid-dev git subversion unixodbc-dev unixodbc autoconf libedit-dev libcurl4-openssl-dev libopus-dev pkg-config libjansson-dev
```

### Downloading Asterisk
```
cd /usr/src
sudo wget http://downloads.asterisk.org/pub/telephony/asterisk/asterisk-22.5.1.tar.gz
sudo tar zxvf asterisk-22.5.1.tar.gz
cd asterisk-22.5.1
```

### Installing per-requisites for asterisk
```
sudo ./contrib/scripts/install_prereq install
sudo ./configure 
```

### After running this you MUST SELECT CODEC OPUS, else you would hear silence even though agent speaks
```
sudo make menuselect
```

### Run these to complete the installation process
```
sudo make -j$(nproc)

sudo make install
sudo make samples
sudo make config
sudo ldconfig
```

### Recommended to keep Asterisk running in background
```
sudo systemctl enable asterisk
sudo systemctl restart asterisk
sudo systemctl status asterisk
```

### Sanity check that asterisk is installed and so is codec opus
```
sudo asterisk -rvvvvvvvvvvvvvvvvvv
module show like opus #run this command in asterisk_cli>
```

# 2. Copying the configuration files to your Asterisk
You can safely copy these files and they will work just fine. You just need to change the 
- ip and local net based on your vm
- sip uri from livekit
- username/password for your zoiper testing

**NOTE: YOU MUST RESTART ASTERISK AFTER COPYING THESE CONFIG FILES, SO THESE CHANGES CAN TAKE PLACE**
Or just do this in asterisk_cli>
```
pjsip reload
dialplan reload

sudo systemctl restart asterisk #or just run this to restart asterisk
```

### pjsip.conf
```
[general]
static=yes
writeprotect=no
autofallthrough=yes

; ---------------------------
; Transports
; ---------------------------
[transport-udp]
type=transport
protocol=udp
bind=0.0.0.0:5063
local_net=LOCAL_NET
external_signaling_address=PUBLIC_IP
external_media_address=PUBLIC_IP

[transport-tcp]
type=transport
protocol=tcp
bind=0.0.0.0:5060
local_net=LOCAL_NET
external_signaling_address=PUBLIC_IP
external_media_address=PUBLIC_IP

; ---------------------------
; LOCAL PHONE (Zoiper)
; ---------------------------
[auth_local]
type=auth
auth_type=userpass
username=LOCAL_USER
password=LOCAL_PASS

[LOCAL_USER]
type=aor
max_contacts=100
remove_existing=yes

[LOCAL_USER]
type=endpoint
context=from-internal
disallow=all
allow=opus,ulaw
aors=LOCAL_USER
auth=auth_local
direct_media=no
rtp_symmetric=yes
force_rport=yes
rewrite_contact=yes
; don’t hardcode transport — let Zoiper pick UDP/TCP
; transport=transport-udp

[identify-LOCAL_USER]
type=identify
endpoint=LOCAL_USER
match=0.0.0.0/0

; ---------------------------
; LIVEKIT TRUNK (use as outbound trunk)
; ---------------------------
[livekit-outbound-auth]
type=auth
auth_type=userpass
username=TRUNK_USER
password=TRUNK_PASS

[registration_livekit]
type=registration
outbound_auth=livekit-outbound-auth
server_uri=sip:LIVEKIT_TENANT.sip.livekit.cloud
client_uri=sip:TRUNK_USER@LIVEKIT_TENANT.sip.livekit.cloud
contact_user=TRUNK_USER
retry_interval=60

[livekit-cloud-aor]
type=aor
; no static contact — we will use outbound_proxy/registration to reach LiveKit

[livekit-cloud]
type=endpoint
context=from-livekit
disallow=all
allow=opus,ulaw
outbound_auth=livekit-outbound-auth
transport=transport-tcp
direct_media=no
rtp_symmetric=yes
force_rport=yes
rewrite_contact=yes
outbound_proxy=sip:LIVEKIT_TENANT.sip.livekit.cloud
media_address=PUBLIC_IP
```

### extensions.conf 
With this any number called on zoiper will be redirected to your number on inbound-trunk (so you may tinker with it any way you like)
```
[general]
static=yes
writeprotect=no
autofallthrough=yes

[from-internal]
; Any numeric dialed from LOCAL_USER will go to LiveKit trunk
exten => _X.,1,NoOp(Forwarding ${EXTEN} to LiveKit via endpoint livekit-cloud)
 same => n,Set(LIVEKIT_NUMBER=DEFAULT_OUTBOUND_NUMBER)
 same => n,NoOp(Calling LiveKit number ${LIVEKIT_NUMBER} for dialed ${EXTEN})
 ; Dial using endpoint livekit-cloud and explicit SIP URI resource
 same => n,Dial(PJSIP/livekit-cloud/sip:${LIVEKIT_NUMBER}@LIVEKIT_TENANT.sip.livekit.cloud,60)
 same => n,Hangup()

[from-livekit]
; Incoming calls from LiveKit will land here
exten => _X.,1,NoOp(Inbound call from LiveKit: ${CALLERID(num)})
 same => n,Answer()
 same => n,Playback(hello-world)   ; replace with your desired handling (AGI, queue, bridge)
 same => n,Hangup()
```

### rtp.conf
```
[general]
rtpstart=10000
rtpend=20000
```

# 3. Opening essential ports so asterisk and livekit can communicate properly
Now we are done with asterisk just open these ports on your cloud provider
```
5060-5070 ANY
10000-20000 UDP
50000-60000 UDP
7881, 7882, 5349, 3478, 4569  ANY
``` 

# 4. Creating inbound-trunk and dispatch rule on Livekit

### Inbound trunk
Under telephony > configuration >  Create New > trunk. 
It should look something like this under JsonEditor, add any number for testing or one you have bought. You may copy this
```
{
  "sipTrunkId": "ST_uMq2BB3o388t",
  "name": "Asterisk-Trunk",
  "numbers": [
    "+15105550100"
  ],
  "krispEnabled": true
}
```

### Dispatch rule
Under telephony > configuration >  Create New > Dispatch rule
You can copy this if you like. 
- For each caller it will create a new room
- You must put same agent name here and in your agent.py
```
{
  "sipDispatchRuleId": "SDR_CbLio6wTAXMe",
  "rule": {
    "dispatchRuleIndividual": {
      "roomPrefix": "call-"
    }
  },
  "name": "Asterisk Dispatch Rule",
  "roomConfig": {
    "agents": [
      {
        "agentName": "asterisk-agent",
        "metadata": "Asterisk inbound routing"
      }
    ]
  }
}
```

# 5. Simple agent code, so it joins the room when you call
Agent.py (i am using real-time api, change it to STT-LLM-TTS if you like)
```
import logging
from dotenv import load_dotenv
import asyncio
from livekit import agents
from livekit.agents import AgentSession, Agent, RoomInputOptions, AutoSubscribe, metrics
from livekit.plugins import openai, noise_cancellation

load_dotenv()
logger = logging.getLogger("voice-agent")
logging.basicConfig(level=logging.INFO, format="%(levelname)s:%(name)s:%(message)s")


class Assistant(Agent):
    def __init__(self) -> None:
        super().__init__(
            instructions="You are a helpful AI assistant that MUST always reply in English, no matter what language the user speaks."
        )


async def entrypoint(ctx: agents.JobContext):
    logger.info(f"🔌 Connecting to room {ctx.room.name}")

    # Wait for first participant
    await ctx.connect(auto_subscribe=AutoSubscribe.AUDIO_ONLY)
    participant = await ctx.wait_for_participant()
    logger.info(f"🎤 Participant joined: {participant.identity}")

    # Collect metrics
    usage_collector = metrics.UsageCollector()

    def on_metrics_collected(agent_metrics: metrics.AgentMetrics):
        logger.info("📊 Usage metrics collected")
        metrics.log_metrics(agent_metrics)
        usage_collector.collect(agent_metrics)

    # Setup real-time session
    session = AgentSession(
        llm=openai.realtime.RealtimeModel(voice="coral"),  # streaming LLM with voice output
    )
    session.on("metrics_collected", on_metrics_collected)

    # Log when reply is generated
    @session.on("reply_generated")
    def _on_reply_generated(reply):
        logger.info(f"🤖 Reply generated: {reply.text[:80]}...")

    # Log when reply is sent (played back into room)
    @session.on("reply_played")
    def _on_reply_played(reply):
        logger.info(f"🔊 Reply sent to participant: {reply.text[:80]}...")

    await session.start(
        room=ctx.room,
        agent=Assistant(),
        room_input_options=RoomInputOptions(
            # Enable LiveKit Cloud noise cancellation
            noise_cancellation=noise_cancellation.BVC(),
        ),
    )

    # Immediately greet participant via voice
    greeting = "Boss Doraemon, Welcome, glad to have you here."
    logger.info(f"🤖 Generating greeting reply: {greeting}")
    await session.generate_reply(instructions=greeting)
    logger.info("🔊 Greeting reply sent ✅")


if __name__ == "__main__":
    agents.cli.run_app(
        agents.WorkerOptions(entrypoint_fnc=entrypoint, agent_name="asterisk-agent")
    )
```

# 6. Some common issues you might fall into

**ISSUE 1: You didnt follow my instructions and you cant hear your agent speaking**
Install codec opus from menuselect of asterisk
Also make sure you have opus in your configuration files like above

**ISSUE 2: You done everything above but you cant recieve TCP/UDP packets from livekit OR Zoiper shows ringing only** 
Your government isn't sigma, they have banned those communication ports.You would need vpn when testing with zoiper or buy a number from twilio and connect that

**ISSUE 3: You are using VPN and still zoiper keeps ringing or no packets from livekit are comming** 
You need to setup port-forwarding on your VPN, if you have copy pasted above files, forward port 5061 to 5061 and connect zoiper using 5061 too. Or just buy a twilio number

# 7. Shameless plug to hire me
I am about to graduate in few months and I would love if someone reading this guide would hire me 🥺👉👈. Not good at networking and this was the only place I could think of for shameless plug 😞
You can see my portfolio here: https://attiquers.vercel.app/
 
If i have missed anything let me know.
I hope this fixes your issues, if not you can reach me out too and I would love to help.

**Comment by singh-harveer** [comment]
@attiquers  thank you so much, it solve my issue `opus` was culprit for me. `opus` was missing in my installation.

all configration is okay excep below one
because livekit does not suppoort register method.

```
[livekit-outbound-auth]
  type=auth
  auth_type=userpass
  username=TRUNK_USER
  password=TRUNK_PASS
  
  [registration_livekit]
  type=registration
  outbound_auth=livekit-outbound-auth
  server_uri=sip:LIVEKIT_TENANT.sip.livekit.cloud
  client_uri=sip:TRUNK_USER@LIVEKIT_TENANT.sip.livekit.cloud
  contact_user=TRUNK_USER
  retry_interval=60
```




Error from livekit when tried with registeration and password based authentication.

`res_pjsip_outbound_registration.c:1550 handle_registration_response: Fatal response '405' received from 'sip:tenatID.sip.livekit.cloud' on registration attempt to 'sip:user_name@tenatID.sip.livekit.cloud', stopping outbound registration
`

I am still testing few thing to find correct config, once i find stable config for pjsip and rtp i will share here.

About hiring I will definitely recommend you if I come accross any opening :)

thanks again!

**Comment by attiquers** [comment]
I'm happy that I can help. 

You may be right about registration since livekit uses SIP URI. I only use asterisk for testing with zoiper and then buy twillio numbers for clients when deploying.

I appreciate you recommending me for job openings, thank you 😁

**Comment by erwiower2** [comment]
But please be careful if you use chan_sip and open port 5060 to the internet. That is very insecure and can lead to a very large phone bill!

At first, I was very enthusiastic about Livekit, but I no longer use it for the following reasons: I had to write a separate script to monitor whether I put the call on hold and then artificially forward the participant to another extension so that my MOH wouldn’t sound completely choppy due to Livekit’s noise cancellation! However, I had extreme problems in a call flow when I had an active participant, wanted to put them on hold, and then forward them with consult.

Imagine:

Me (Livekit Noise Cancellation) → Participant Outbound → Answered → Consult forward (Hold on Participant Outbound) → Answered → Transfer → Call drops

The call drops because there were issues in the dialplan logic. I had to query the call status, because otherwise all channels would be terminated when a hold occurred. Therefore, I had to set the DB value for callstatus to hold during a hold so that the Outbound participant would NOT be hung up.

**Comment by nksaqlain** [comment]
@singh-harveer following does the job easily. Make sure to define this in `pjsip_wizard.conf` instead of `pjsip.conf`

```
[livekit-sip]
type = wizard
sends_auth = no
accepts_auth = no
sends_registrations = no
accepts-registrations = no
remote_hosts = LIVEKIT_TENANT.sip.livekit.cloud
endpoint/context = default
aor/qualify_frequency = 15
```


It is similar to using IP based SIP Trunk in conventional chan_sip without the need of providing credentials.

**Comment by acharan-tech-200037** [comment]
In my case the livekit agent is connected , but there is no sound and also no messages were going to agent from phone 

Logs are: 
Connected to Asterisk 23.2.2 currently running on HL1 (pid = 185589)
    -- Executing [999@from-internal:1] Answer("PJSIP/1001-00000000", "") in new stack
    -- Executing [999@from-internal:2] Playback("PJSIP/1001-00000000", "demo-congrats") in new stack
    -- <PJSIP/1001-00000000> Playing 'demo-congrats.gsm' (language 'en')
  == Spawn extension (from-internal, 999, 2) exited non-zero on 'PJSIP/1001-00000000'
    -- Executing [1000@from-internal:1] NoOp("PJSIP/1001-00000001", "Forwarding 1000 to LiveKit via endpoint livekit-cloud") in new stack
    -- Executing [1000@from-internal:2] Set("PJSIP/1001-00000001", "LIVEKIT_NUMBER=+14849622010") in new stack
    -- Executing [1000@from-internal:3] NoOp("PJSIP/1001-00000001", "Calling LiveKit number +14849622010 for dialed 1000") in new stack
    -- Executing [1000@from-internal:4] Dial("PJSIP/1001-00000001", "PJSIP/livekit-cloud/sip:+14849622010@6bqbis3uqij.sip.livekit.cloud,60") in new stack
    -- Called PJSIP/livekit-cloud/sip:+14849622010@6bqbis3uqij.sip.livekit.cloud
    -- PJSIP/livekit-cloud-00000002 is ringing
    -- PJSIP/livekit-cloud-00000002 is ringing
    -- PJSIP/livekit-cloud-00000002 is ringing
    -- PJSIP/livekit-cloud-00000002 answered PJSIP/1001-00000001
    -- Channel PJSIP/livekit-cloud-00000002 joined 'simple_bridge' basic-bridge <c350eb3e-d2b5-41d0-af3c-509a62d4e69d>
    -- Channel PJSIP/1001-00000001 joined 'simple_bridge' basic-bridge <c350eb3e-d2b5-41d0-af3c-509a62d4e69d>

**Comment by nishadmusthafa** [comment]
Hi, I would request that you use https://community.livekit.io/ or email livekit support to get help debugging issues that are specific to your setup. Given this issue has been mostly due to individual setups and that it's been a while, I'm going to close this one out.
